### ISSUE TITLE
Rule proposal: simplify negation

### ISSUE THREAD

**Issue by noftaly** [sindresorhus/eslint-plugin-unicorn#856]
## Description
This rule would enforce users to simplify negations: `!(a === b)` -> `a !== b`.
There might be other conditions like this that could be simplified.
I don't know about the rule's name though, `simplify-negation` might be too vague...

## Fail

```js
if (!(a > b)) {}
const foo = !(bar === baz);
const foo = !(typeof bar === 'undefined');
```

## Pass

```js
if (a <= b) {}
const foo = bar !== baz;
const foo = typeof bar !== 'undefined';
const foo = !Array.isArray(bar);
```

**Comment by fisker** [comment]
I love this one, it will be great if we can convert big nested `LogicExpression`.

**Comment by papb** [comment]
"Fun" fact: `!(a > b)` is not strictly equal to `(a <= b)`...

```js
console.log(!(null > undefined)); // true
console.log(null <= undefined); // false
```

**Comment by yakov116** [comment]
What about 2 conditions?

```js
	if (!(a===b && c === d)) {}
```

Sometimes that way is easier to read

**Comment by papb** [comment]
> Sometimes that way is easier to read

I agree; some times the non-[DeMorgan](https://en.wikipedia.org/wiki/De_Morgan%27s_laws) way is more readable

**Comment by noftaly** [comment]
Then maybe add an option to not optimize when there are multiple ANDs, or nested ANDs and ORs? Or we also just could not add an option and letting the user disable the rule inline.

Concerning the implementation, we could use some algebra expressions optimization algorithms, and check if it is shorter than the original one, and if it is, replace? This will only work with boolean operators tho, not `>`.
But do we really want to support that? As @papb, fixing it might break the code

**Comment by sindresorhus** [comment]
> if (!(a===b && c === d)) {}

I much prefer this too. If there's going to be an option, it should prefer this by default.

**Comment by sindresorhus** [comment]
> But do we really want to support that? As @papb said, fixing it might break the code

Yes, but they should not be auto-fixable. We can make them use the suggestions API instead.

**Comment by sindresorhus** [comment]
> we could use some boolean-expressions optimization algorithms

Do you have any in mind?

**Comment by noftaly** [comment]
> Do you have any in mind?

Not really, and I found online and did not found any in JS. But I was thinking of something like this: [boolean-algebra](http://boolean-algebra.com/)
It will, for example, simplify an expression like this 
![image](https://user-images.githubusercontent.com/34779161/101784763-3927a080-3afc-11eb-8bf0-ab3d30d44379.png)
to this : `!A + !B`

**Comment by sindresorhus** [comment]
This is now accepted.

It should prefer readability by default, for example: `if (!(a === b && c === d)) {}`.

**Comment by fisker** [comment]
Similar rule  https://github.com/SonarSource/eslint-plugin-sonarjs/blob/master/docs/rules/no-inverted-boolean-check.md

**Comment by fregante** [comment]
> this: [boolean-algebra](http://boolean-algebra.com/)
> It will, for example, simplify an expression like this

I was thinking about a rule that did this exactly: `simplify-boolean-algebra` or something like that.

Instead of manually handling each boolean-related awkwardness, why not just simplify whatever expression we see? I regularly need this, however I need it across if/else/elseif/return rather than in one `if (expression)`

**Comment by jguddas** [comment]
The code is not in a presentable state, but it works 🎉 

https://astexplorer.net/#/gist/de69908bb1794506afe642b14811684f/ff8b40171526ba98cea4ff3c5bebe8a50440b375

Please let me know when you find something that does not work as you expect and generally what you have tried out, so we can include that in our test cases.

**Comment by jguddas** [comment]
What about binary expressions like `a - -b` which can be simplified down to `a + b`?

**Comment by noftaly** [comment]
You have to be careful simplifying expressions with `-` and `+` because of automatic type inference. There are edge cases where `a - -b` is not the same as `a + b`
```sh
> 3 - -[]
3
> 3 + []
'3'
> 3 - -{}
NaN
> 3 + {}
'3[object Object]'
> 3 - -Number
NaN
> 3 + Number
'3function Number() { [native code] }'
```

**Comment by jguddas** [comment]
You are right, what about fixing `3 - -{}` to `{} + 3`, that would work?

**Comment by fisker** [comment]
There is a rust version https://github.com/rome/tools/blob/22dd11d5d03486d3b6f6679f398cf0bbc440a2ef/crates/rome_js_analyze/src/analyzers/complexity/use_simplified_logic_expression.rs#L216

**Comment by yukulele** [comment]
```js
if (!(a === b && c === d)) {}
```
become
```js
if(a !== b || c !== d) {}
```

**Comment by jguddas** [comment]
@yukulele my solution was to add a `reduceParens` option that is disabled by default so you would only get that case if you want it to, by default it would stay `!(a === b && c === d)`.

**Comment by yusufkandemir** [comment]
Recently came upon this: https://github.com/azat-io/eslint-plugin-de-morgan
